Spline Fixes V2

danieldresser-ie opened the pull request:
Maybe Travis will run this time.

danieldresser-ie commented:
Hmm, starting a new PR got travis to run the tests.  Which turned up a complaint about signed/unsigned comparison, but after fixing that, everything is passing on Travis.  So I'm not sure why it's failing for you.

Could you check whether it is the newly added linear test which is failing, or is it one of the other curve types which was already tested?

johnhaddon commented:
> Could you check whether it is the newly added linear test which is failing, or is it one of the other curve types which was already tested?

I put my previous comment on the actual line of the test that crashed - it was the linear integration test. I'll try again here...

danieldresser-ie commented:
Well, that took me longer than it should have to spot.  At first I was looking at the actual solve() method, rather than the wrapper that also fills in the 4 Y values.  That wrapper had a straightforward bug ( it hadn't been updated to only read up to coefficientsNeeded, and was instead hardcoded to read 4 Y values, which took it past the end of the array ).

The frustrating thing is that back in mid-May when I took the first stab at this, I had written a note "make sure everything works with less than 4 coefficients needed", and once I realized what was going on, I'm pretty sure I had identified this exact issue at the time.  But by the time I got back to it, I'd forgotten what issue I was thinking about, and I couldn't spot any problems, so I just added the test and then concluded that it must be fine since the test passed.

Oh well, all seems good now, and I'm pretty sure I've fixed the crash you saw.

johnhaddon commented:
Yep, that fixes it for me - nice one.
